jscomp/lib/dom.ml

```diff
@@ -0,0 +1,295 @@
+and element = object
+  inherit node
+
+  method tagName : string
+  method tagName__ : string
+
+  method getAttribute : string -> string opt 
+  method getAttribute__1 : string -> string opt 
+
+  method setAttribute : string -> string -> unit 
+  method setAttribute__2 : string -> string -> unit 
+
+  method removeAttribute : string -> unit 
+  method removeAttribute__1 : string -> unit 
+
+  method hasAttribute : string -> boolean  
+  method hasAttribute__1 : string -> boolean  
+
+  method getAttributeNS : string -> string -> string opt 
+  method getAttributeNS__2 : string -> string -> string opt 
+
+  method setAttributeNS : string -> string -> string -> unit 
+  method setAttributeNS__3 : string -> string -> string -> unit 
+
+  method removeAttributeNS : string -> string -> unit 
+  method removeAttributeNS__2 : string -> string -> unit 
+
+  method hasAttributeNS : string -> string -> boolean  
+  method hasAttributeNS__2 : string -> string -> boolean  
+
+  method getAttributeNode : string -> attr  opt 
+  method getAttributeNode__1 : string -> attr  opt 
+
+  method setAttributeNode : attr  -> attr  opt 
+  method setAttributeNode__1 : attr  -> attr  opt 
+
+  method removeAttributeNode : attr  -> attr  
+  method removeAttributeNode__1 : attr  -> attr  
+
+  method getAttributeNodeNS : string -> string -> attr  opt 
+  method getAttributeNodeNS__2 : string -> string -> attr  opt 
+
+  method setAttributeNodeNS : attr  -> attr  opt 
+  method setAttributeNodeNS__1 : attr  -> attr  opt 
+
+  method getElementsByTagName : string -> element nodeList  
+  method getElementsByTagName__1 : string -> element nodeList  
+
+  method attributes : attr namedNodeMap  
+  method attributes__ : attr namedNodeMap
+end
+
+class type characterData = object
+  inherit node
+
+
+  method data__get : string 
+  method data__set : string 
+
+  method length : int
+  method length__ : int
+
+  method subjs_stringData : int -> int -> string 
+  method subjs_stringData__2 : int -> int -> string 
+
+  method appendData : string -> unit 
+  method appendData__1 : string -> unit 
+
+  method insertData : int -> string -> unit 
+  method insertData__2 : int -> string -> unit 
+
+  method deleteData : int -> int -> unit 
+  method deleteData__2 : int -> int -> unit 
+
+  method replaceData : int -> int -> string -> unit 
+  method replaceData__3 : int -> int -> string -> unit 
+end
+
+class type comment = characterData
+
+class type text = characterData
+
+class type documentFragment = node
+
+class type ['element] document = object
+  inherit node
+  method documentElement : 'element  
+  method documentElement__ : 'element  
+
+  method createDocumentFragment__ : documentFragment  
+  (* [Function.prototype.length] is 0*)
+
+  method createElement : string -> 'element  
+  method createElement__1 : string -> 'element  
+
+  method createElementNS : string -> string -> 'element  
+  method createElementNS__2 : string -> string -> 'element  
+
```

Comment:
this is too verbose and error prone, 
for `4.02` we can only use `class type xx = object end [@@js.class_type]` to save typing..